### PR TITLE
LWG Poll 2: P2051R0 C++ Standard Library Issues to be moved in Prague

### LINKED ISSUE #3704: [2020-02 LWG Poll 2] P2051R0 C++ Standard Library Issues to be moved in Prague
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2051r0.html

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #3704.

Notes/issues:
* LWG3281 not applied; text was removed by LWG3282

Fixes cplusplus/papers#781.
Fixes cplusplus/nbballot#298.
Fixes cplusplus/nbballot#258.
Fixes cplusplus/nbballot#154.
Fixes cplusplus/nbballot#300.
Fixes cplusplus/nbballot#283.
Fixes cplusplus/nbballot#282.
Fixes cplusplus/nbballot#214.
Fixes cplusplus/nbballot#215.
Fixes cplusplus/nbballot#216.
Fixes cplusplus/nbballot#290.
Fixes cplusplus/nbballot#160.
Fixes cplusplus/nbballot#161.
Fixes cplusplus/nbballot#329.
Fixes cplusplus/nbballot#330.
Fixes cplusplus/nbballot#331.
Fixes cplusplus/nbballot#340.
Fixes cplusplus/nbballot#243.
Fixes cplusplus/nbballot#210.
Fixes cplusplus/nbballot#295.
Fixes cplusplus/nbballot#176.
Fixes cplusplus/nbballot#299.
Fixes cplusplus/nbballot#278.
Fixes cplusplus/nbballot#222.
Fixes cplusplus/nbballot#199.
Fixes cplusplus/nbballot#179.
Fixes cplusplus/nbballot#200.

### CODE REVIEW

**Comment by jwakely** [comment]
Is everything in this branch also in  motions-2020-02-lwg-3 ?

File: source/time.tex [hunk 1]
```diff
@@ -9372,31 +9438,32 @@
...
-      -> zoned_time<common_type_t<Duration, seconds>>;
-
-  template<class Duration, class TimeZonePtr, class TimeZonePtr2>
-    zoned_time(TimeZonePtr, zoned_time<Duration, TimeZonePtr2>, choose = choose::earliest)
-      -> zoned_time<common_type_t<Duration, seconds>, TimeZonePtr>;
+  template<class TimeZonePtrOrName>
+    using @\exposid{time-zone-representation}@ =
+      conditional_t<is_convertible_v<TimeZonePtrOrName, string_view>,
+                    const time_zone*,
+                    remove_cv_ref<TimeZonePtrOrName>>;          // \expos
```
Comment (johelegp) [inline source/time.tex]: Other multi-line declarations have `\expos` on the line with the exposition-only identifier being declared.

File: source/utilities.tex [hunk 2]
```diff
@@ -7607,8 +7616,9 @@
 \effects
 Equivalent to:
 \begin{codeblock}
-return ::new(static_cast<void*>(p))
-  T(make_obj_using_allocator<T>(alloc, std::forward<Args>(args)...));
+return apply([&](auto&&...xs) {
+         return construct_at(p, std::forward<decltype(xs)>(xs)...);
```
Comment (johelegp) [inline source/utilities.tex]: Can't we editorially use `[&]<class... U>(U&&.. xs) {` and `std::forward<U>(xs)...` here?

**Comment by johelegp** [comment]
> Is everything in this branch also in motions-2020-02-lwg-3 ?

Here's a diff: https://github.com/cplusplus/draft/compare/db3a6049bd7ae1244d0c6a09cfd3ea4b69cdf452..ea509911557a236702dccba48f2d2a2fa55322d0.

File: source/containers.tex [hunk 3]
```diff
@@ -10514,6 +10525,9 @@
   template<class ElementType, size_t Extent = dynamic_extent>
     class span;
 
+  template<class ElementType, size_t Extent>
+    inline constexpr bool ranges::enable_view<span<ElementType, Extent>> =
+      Extent == 0 || Extent == dynamic_extent;
```
[the remark this answers is not in the document]
Comment (timsong-cpp) [reply source/containers.tex]: `span` of nonzero fixed extent are not default constructible and therefore not `view`s.

[the remark this answers is not in the document]
Comment (johelegp) [reply source/containers.tex]: That's right. Thank you.

**Comment by jensmaurer** [comment]
LWG 3280 also fixes US 302, but is not marked as such.

**Comment by burblebee** [comment]
> Is everything in this branch also in motions-2020-02-lwg-3 ?

Yes, lwg-3 was based on lwg-2 in order to resolve conflicts where issues from both motions changed the same wording.

**Comment by jensmaurer** [comment]
LWG 3281 also fixes US 284, but is not marked as such.

File: source/utilities.tex (hunk 2 above)
Comment (burblebee) [reply source/utilities.tex]: I made this change in a separate edit - will leave it up to @zygoloid to see if he wants to include it as part of the motion or not.

File: source/containers.tex [hunk 4]
```diff
@@ -11122,20 +11130,19 @@
 \indexlibrarymember{const_iterator}{span}%
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex]: Thanks!

**Comment by burblebee** [comment]
> LWG 3280 also fixes US 302, but is not marked as such.
> ....

Thanks for pointing these out!  Done  - please recheck.

**Comment by burblebee** [comment]
> > LWG 3280 also fixes US 302, but is not marked as such.
> > ....
> 
> Thanks for pointing these out! Done - please recheck.

Ugh - I did force push as Jens suggested, but that was probably a mistake - it will screw things up for @zygoloid when he tries to merge motions-2020-02-lwg-3 which was based on this :(.
@zygoloid, should I force-push motions-2020-02-lwg-3 also?

File: source/containers.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (johelegp) [reply source/containers.tex]: > [`span::span()` has the operands to `||` in reverse order.](https://github.com/cplusplus/draft/pull/3749#discussion_r381765731)

There are more occurrences. The preference seems to be to have ` == dynamic_extent` first any `|| Extent ==` second, which I second here too.

File: source/utilities.tex [hunk 5]
```diff
@@ -7607,8 +7616,9 @@
 \effects
 Equivalent to:
 \begin{codeblock}
-return ::new(static_cast<void*>(p))
-  T(make_obj_using_allocator<T>(alloc, std::forward<Args>(args)...));
+return apply([&]<class... U>(U&&...xs) {
```
Comment (johelegp) [inline source/utilities.tex]:
```suggestion
return apply([&]<class... U>(U&&... xs) {
```

**Comment by burblebee** [comment]
> > > LWG 3280 also fixes US 302, but is not marked as such.
> > > ....
> > 
> > 
> > Thanks for pointing these out! Done - please recheck.
> 
> Ugh - I did force push as Jens suggested, but that was probably a mistake - it will screw things up for @zygoloid when he tries to merge motions-2020-02-lwg-3 which was based on this :(.
> @zygoloid, should I force-push motions-2020-02-lwg-3 also?

I'm going to go ahead and force push motions-2020-02-lwg-3 rebased on motions-2020-02-lwg-2, hope that's OK.

**Comment by johelegp** [comment]
The [force-pushed](https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095) link shows that one of the changes was lost.

**Comment by burblebee** [comment]
> The [force-pushed](https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095) link shows that one of the changes was lost.

Oh?  Which one?

**Comment by johelegp** [comment]
> > The [force-pushed](https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095) link shows that one of the changes was lost.
> 
> Oh? Which one?

https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095#diff-0754a9e20af4b80758260aaad60a4d1eL3100

**Comment by burblebee** [comment]
> > > The [force-pushed](https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095) link shows that one of the changes was lost.
> > 
> > 
> > Oh? Which one?
> 
> https://github.com/cplusplus/draft/compare/46b038136e41f02ddeac66882ff53c5dff42a849..4185912a94c33ca909de86b59e0167b48a00b095#diff-0754a9e20af4b80758260aaad60a4d1eL3100

That was intentional, see my modified comment above (I guess we added comments here at the same time :) )

**Pull request merged**